### ISSUE TITLE
allow MacOS OpenGL software  renderer?

### ISSUE THREAD

**Issue by jjYBdx4IL** [glfw/glfw#2080]
Seems to work by simply replacing:

> ADD_ATTRIB(NSOpenGLPFAAccelerated);

with

> ADD_ATTRIB(NSOpenGLPFARendererID);ADD_ATTRIB(kCGLRendererGenericFloatID);

in src/nsgl_context.m.

An automatic fallback would be the best option.

**Comment by Khrysys** [comment]
Was it throwing a specific error?

**Comment by jjYBdx4IL** [comment]
> error (65545): NSGL: Failed to find a suitable pixel format

when calling `initWithAttributes`.

**Comment by iTrooz** [comment]
I agree that this would be a ggreat feature, I need it for the same reason (testing an app on a MacOS virtual machine)

**Comment by ikus060** [comment]
Is that something that is working today ? Is it possible to of this project in order to run openGL application on MacOS running within a VM ?

**Comment by iTrooz** [comment]
Idk on upstream, but if you check my linked PR (https://github.com/WerWolv/ImHex/pull/708) we've been using it in ImHex for a while now

**Comment by HinTak** [comment]
I can answer the last question - when the software renderer is in use, `glGetString(GL_RENDERER)` after obtaining a valid GL context, reports itself as `b'Apple Software Renderer' `. On genuine Mac os desktops, we have seen `b'Apple M3 Pro'` and `b'Apple M1'` for that field (reported by some skia python users in one of our mac os x specific issues, 214) . So you can read that field to see what actual hardware/software is in use. There is probably a more convenient way.

**Comment by HinTak** [comment]
My interests in this is github's own CI testing (which is headless), and I have no apple hardware nor VM. Looking at the code, I think it is possible to add automatic fall-back? The is no simple way of doing `DELETE_ATTRIB`, but it seems to be possible to do a `REPLACE_ATTRIB`. Slight complications of replacing 1 with 2 - I guess if we are sure it was added earlier, we can everything upward by 1, then insert the two in the middle overwriting the previous one.

**Comment by HinTak** [comment]
I have added automatic software renderer retry if the accelerated request fails: https://github.com/glfw/glfw/pull/2571 - can somebody with genuine mac hardware gives it a try to check that accelerated opengl still works?

**Comment by ColleagueRiley** [comment]
This is how RGFW supports this feature: 

https://github.com/ColleagueRiley/RGFW/blob/d80c8bea40bb1b367941d9b119bbac9f2fa329c3/RGFW.h#L7467

**Comment by iTrooz** [comment]
> I have added automatic software renderer retry if the accelerated request fails: #2571 - can somebody with genuine mac hardware gives it a try to check that accelerated opengl still works?

Sorry, did not see this earlier.
How can you know if you are using accelerated opengl or not, once the window has opened ?

**Comment by HinTak** [comment]
> > I have added automatic software renderer retry if the accelerated request fails: #2571 - can somebody with genuine mac hardware gives it a try to check that accelerated opengl still works?
> 
> Sorry, did not see this earlier. How can you know if you are using accelerated opengl or not, once the window has opened ?

See https://github.com/glfw/glfw/issues/2080#issuecomment-2145902645

**Comment by iTrooz** [comment]
> I have added automatic software renderer retry if the accelerated request fails: #2571 - can somebody with genuine mac hardware gives it a try to check that accelerated opengl still works?

My friend just confirmed that: https://github.com/glfw/glfw/issues/2080#issuecomment-2146431760

**Comment by rafael2k** [comment]
We hit this running a Fyne (Go, via `go-gl/glfw` v3.4) app inside a GPU-less
QEMU/KVM macOS guest (macOS 15.7.7), and I wanted to share a minimal fix plus
the evidence that a usable software renderer really is available — it's just
never requested.

**A software renderer exists; NSGL just isn't asked for it.** A
`CGLQueryRendererInfo()` probe on the guest reports exactly one renderer:

```
renderer count: 1
  [0] id=0x01020400 accelerated=0 online=1 window=1
```

`0x01020400 & kCGLRendererIDMatchingMask (0x00FE7F00) == 0x00020400 ==
kCGLRendererGenericFloatID` — the Apple Software Renderer, `accelerated=0`,
`window=1` (can back a window). But `_glfwCreateContextNSGL()` adds
`NSOpenGLPFAAccelerated` as `attribs[0]`, which *requires* a hardware renderer,
so `initWithAttributes:` returns `nil` and creation fails with
`NSGL: Failed to find a suitable pixel format`.

**Minimal fix — retry once without the accelerated attribute.** If the
accelerated pixel format is unavailable, retry from `attribs + 1` (dropping only
`NSOpenGLPFAAccelerated`, which is always first), letting macOS return the
software renderer instead of failing:

```objc
    window->context.nsgl.pixelFormat =
        [[NSOpenGLPixelFormat alloc] initWithAttributes:attribs];
    if (window->context.nsgl.pixelFormat == nil)
    {
        // GPU-less fallback: drop attribs[0] (NSOpenGLPFAAccelerated) and retry
        // so macOS can return its software (GenericFloat) renderer.
        assert(attribs[0] == NSOpenGLPFAAccelerated);
        window->context.nsgl.pixelFormat =
            [[NSOpenGLPixelFormat alloc] initWithAttributes:(attribs + 1)];
    }
    if (window->context.nsgl.pixelFormat == nil)
    {
        _glfwInputError(GLFW_FORMAT_UNAVAILABLE,
                        "NSGL: Failed to find a suitable pixel format");
        return GLFW_FALSE;
    }
```

**Why this is safe on hardware:** the accelerated request runs first and
succeeds on any real GPU, so that path is byte-for-byte unchanged and the
fallback is never reached. The fallback only triggers when context creation
would *otherwise fail entirely*, so software rendering is strictly better than
the current hard failure. It's also non-Cocoa-neutral (`nsgl_context.m` is
macOS-only).

Verified: with this change a Fyne app (requesting legacy GL 2.1) opens and
renders — software, so sluggish, but fully functional — in the GPU-less guest;
no change on real Macs.

Happy to turn this into a PR. If you'd prefer it gated behind a hint rather than
an unconditional fallback (per the request in this issue), that's an easy
variant — the position-independent version would strip `NSOpenGLPFAAccelerated`
from the attribute list when the hint is set instead of relying on `attribs[0]`.

Related, for downstream context: go-gl/glfw#335, fyne-io/fyne#2373.
